linera-core/src/client/mod.rs

```diff
@@ -1144,73 +1146,105 @@ where
+                let block_batch_range = BlockHeightRange::multi(first_block.height, batch_size);
+                let query = ChainInfoQuery::new(chain_id)
+                    .with_sent_certificate_hashes_in_range(block_batch_range.clone());
+
+                let local_response = self.local_query(query.clone()).await?;
+                let highest_found = Self::highest_found(&query, &local_response);
+                let blocks_missing = block_batch_range
+                    .highest_expected()
+                    .saturating_sub(highest_found)
+                    .0;
+
+                new_tracker += batch_size - blocks_missing;
+                if blocks_missing == 0 {
+                    // All certificates for blocks in `block_batch` are found on local node.
+                    // Move to the next batch.
+                    continue;
+                }
 
-            let remote_response = remote_node.handle_chain_info_query(query).await?;
-            let certificate_hash =
-                match remote_response.requested_sent_certificate_hashes.as_slice() {
-                    &[hash] => hash,
-                    [] => {
-                        warn!("Validator didn't have certificate he claimed to have.");
-                        break;
-                    }
-                    _ => {
-                        error!("Validator sent more than one certificate hash for a single block.");
+                let request_missing_in_batch =
+                    BlockHeightRange::multi(highest_found.try_add_one()?, blocks_missing);
+                let remote_query = ChainInfoQuery::new(chain_id)
+                    .with_sent_certificate_hashes_in_range(request_missing_in_batch.clone());
+
+                let remote_response = remote_node.handle_chain_info_query(remote_query).await?;
+
+                let certificate_hashes = match remote_response
+                    .requested_sent_certificate_hashes
+                    .as_slice()
+                {
+                    hashes if hashes.len() as u64 != request_missing_in_batch.limit.unwrap() => {
+                        error!(
+                            first_block_height = request_missing_in_batch.start.0,
+                            requested_num = request_missing_in_batch.limit.unwrap(),
+                            received_num = hashes.len(),
+                            "Validator sent invalid number of certificate hashes."
+                        );
                         return Err(NodeError::InvalidChainInfoResponse);
                     }
+                    hashes => hashes.to_vec(),
                 };
 
-            let certificate = remote_node
-                .node
-                .download_certificate(certificate_hash)
-                .await?;
+                let remote_certificates = remote_node
+                    .node
+                    .download_certificates(certificate_hashes)
+                    .await?;
```

Comment:
In the case of the admin chain there will be lots of sender chains each of which sent only one message, so this will just be one hash. It would be better to batch the hashes from the different sender chains together.

Reply:
But the `ChainInfoQuery` expects the `chain_id` of the chain I'm asking about, so it's not possible to request for certificate hashes for different chains 🤔

Reply:
Ah, I can request a certificate hashes independently but request certificates as a batch – that'd would work.

It's still unoptimal though, that for `N` certificates (from the batch) we need to do `N+1` queries, rather than just `N`